Capture ExecutionContext for Dispatcher.InvokeAsync

walterlv opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->

To make `AsyncLocal<T>` work correctly with `Dispatcher.InvokeAsync`, we need to capture the current `ExecutionContext` and pass it to the dispatcher. This allows the async local state to be preserved across asynchronous calls.

See the code below for implementation details.

```csharp
public MainWindow()
{
    InitializeComponent();
    Loaded += OnLoaded;
}

private readonly AsyncLocal<object?> _service = new AsyncLocal<object?>();

private async void OnLoaded(object? sender, RoutedEventArgs e)
{
    // Test 1:
    _service.Value = new object();
    Dispatcher.UIThread.Invoke(() =>
    {
        var foo = _service.Value;
        Console.WriteLine($"Service value 1: {foo}");
    });

    // Test 2:
    _service.Value = new object();
    Task.Run(() =>
    {
        var foo = _service.Value;
        Console.WriteLine($"Service value 2: {foo}");
    });

    // Test 3:
    _service.Value = new object();
    Dispatcher.UIThread.InvokeAsync(() =>
    {
        var foo = _service.Value;
        Console.WriteLine($"Service value 3: {foo}");
    });
}
```

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->

The code block above generates the following output when run:

```
Service value 1: System.Object
Service value 2: System.Object
Service value 3:
```

The third test returns `null`, which shows that the `AsyncLocal<T>` value is not preserved when using `Dispatcher.InvokeAsync`.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->

The third test should now return the same value as the first two tests, preserving the `AsyncLocal<T>` state across asynchronous calls.

To fix the third test, we can add multiple lines as shown below:

```diff
    _service.Value = new object();
++  var context = ExecutionContext.Capture();
    Dispatcher.UIThread.InvokeAsync(() =>
    {
++      ExecutionContext.Restore(context);
        var foo = _service.Value;
        Console.WriteLine($"Service value: {foo}");
    });
```

WPF works correctly because it captures the `ExecutionContext` in the constructor of `DispatcherOperation` and restores it in the `InvokeAsync` method. I believe we should make Avalonia work the same way as WPF's `DispatcherOperation` and .NET's `Task.Run` method.

- [The `Capture` code in WPF](https://github.com/dotnet/wpf/blob/a57aae82cc2e6693e450186e5216a88a3bb48304/src/Microsoft.DotNet.Wpf/src/WindowsBase/System/Windows/Threading/DispatcherOperation.cs#L38)

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->

I added a new field `ExecutionContext? _executionContext` to the `DispatcherOperation` class and modified the constructor to capture the current `ExecutionContext`. Then, I restored the captured context before executing the `InvokeCore` method.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
<!--- List any breaking changes here. -->

If someone uses `AsyncLocal<T>` to store values and expects `null` in this situation, this PR may break their code. However, this is not a common use case, and the fix is necessary to ensure that `AsyncLocal<T>` works correctly with all other asynchronous methods in .NET.

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

walterlv committed:
Capture ExecutionContext for Dispatcher.InvokeAsync

(cherry picked from commit e41c9272ac45afc31007a8fe26f9f56291b063ef)

walterlv commented:
I've tested the new package and got the expected result:

![The Rider debugger view](https://github.com/user-attachments/assets/00700cad-e2a2-4d31-8849-270b1134badb)

kekekeks commented on src/Avalonia.Base/Threading/DispatcherOperation.cs:
```diff
@@ -52,6 +53,7 @@ private protected DispatcherOperation(Dispatcher dispatcher, DispatcherPriority
         ThrowOnUiThread = throwOnUiThread;
         Priority = priority;
         Dispatcher = dispatcher;
+        _executionContext = ExecutionContext.Capture();
```
WPF checks for ExecutionContext.IsFlowSuppressed() - https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/CulturePreservingExecutionContext.cs#L85

kekekeks commented on src/Avalonia.Base/Threading/DispatcherOperation.cs:
```diff
@@ -103,7 +105,7 @@ public event EventHandler Completed
                 _completed += value;
             }
         }
-        
```
Whitespace changes

kekekeks replied:
Unfortunately whitespace changes pollute `git blame` output.

We are planning to somewhat solve this in one go later, but for now please remove whitespace changes.

kekekeks reviewed (dismissed):
1) this PR doesn't contain the logic from [CulturePreservingExecutionContext](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/CulturePreservingExecutionContext.cs) that

walterlv committed:
Implement CulturePreservingExecutionContext

walterlv committed:
Add IsFlowSuppressed checking

walterlv committed:
Add NET6_0_OR_GREATER because only the Restore need it.

walterlv committed:
Use `ExecutionContext.Run` instead of `ExecutionContext.Restore`.

walterlv committed:
Pass this to avoid lambda capture.

walterlv commented on src/Avalonia.Base/Threading/CulturePreservingExecutionContext.cs:
```diff
@@ -0,0 +1,151 @@
+    /// <param name="executionContext">The execution context to run in.</param>
+    /// <param name="callback">The callback to execute.</param>
+    /// <param name="state">The state to pass to the callback.</param>
+    public static void Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, object? state)
```
Thanks for the suggestion. I don't think making `ContextCallback` generic would be beneficial here. Since `ExecutionContext.Run()` requires the non-generic version, we'd need an extra delegate wrapper that would be slower than the simple cast we currently use.

The current approach follows the same pattern as the framework's `ExecutionContext`, and the added complexity wouldn't provide much value for what's essentially a pass-through parameter.

MrJul commented:
In modern .NET (Core), the cultures are properly captured by the [`ExecutionContext`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.executioncontext?view=net-9.0):
> In .NET Core, the security context and call context are not supported, however, the impersonation context and culture would typically flow with the execution context.

So we can probably make the use of `CulturePreservingExecutionContext` conditional and simply use `ExecutionContext` directly `#if NET6_0_OR_GREATER`, avoid an extra allocation per dispatcher operation.

walterlv committed:
Use ExecutionContext directly on NET6_0_OR_GREATER

walterlv committed:
on NET6_0_OR_GREATER, use Restore so we can get a simple stack trace.

MrJul commented:
Thank you @walterlv!
Could you please convert your test above to a proper unit test, so we can validate the behavior now and not regress it later?

walterlv committed:
Add unit tests.

walterlv commented:
@MrJul Done.

Three tests are added and all passed.

walterlv committed:
All test code must run inside Task.Run to avoid interfering with the test

walterlv committed:
First, test Task.Run to ensure that the preceding validation always passes, serving as a baseline for the subsequent Invoke/InvokeAsync tests.
This way, if a later test fails, we have the .NET framework's baseline behavior for reference.

MrJul reviewed (approved):
LGTM, thank you!
